### PR TITLE
User Activity Feed

### PR DESCRIPTION
**Pull request by inahsanjuan**
- Able to record activity when following other users.
- Will add lesson activities in the future after creating categories & lessons.

### CODE REVIEW

File: app/User.php [hunk 1]
```diff
@@ -79,4 +80,27 @@ public function setAvatar($avatar)
...
+        $activity->subject_type = get_class($subject);
+        $activity->name = $this->getActivityName($event, $subject);
+        $activity->user_id = $this->id;
+
+        $activity->save();
+    }
+
+    public function removeActivity($event, $subject)
+    {
+        $activity = UserActivity::where('name', $this->getActivityName($event, $subject))->where('subject_id', $subject->id)->first();
```
Comment (nguyenthanhtung88) [inline app/User.php]: I think this code should be re-written:
```
UserActivity::where('name', $this->getActivityName($event, $subject))
    ->where('subject_id', $subject->id)
    ->first()
    ->delete();
```

Try to break line into small part, for adding or removing is easier, your code look clearer.
If you write in just one line, it's hard to read and maintain

File: app/UserActivity.php [hunk 2]
```diff
@@ -10,6 +10,23 @@ class UserActivity extends Model
 
     public function user()
     {
-        $this->belongsTo(User::class);
+        return $this->belongsTo(User::class);
     }
+
+    public function subject()
+    {
+        return $this->morphTo();
```
Comment (nguyenthanhtung88) [inline app/UserActivity.php]: `morphTo` what?

File: app/UserActivity.php [hunk 3]
```diff
@@ -10,6 +10,23 @@ class UserActivity extends Model
...
+    }
+
+    public function recordActivity($event, $subject)
+    {
+        $activity = new UserActivity;
+        $activity->subject_id = $subject->id;
+        $activity->subject_type = get_class($subject);
+        $activity->name = $this->getActivityName($event, $subject);
```
Comment (nguyenthanhtung88) [inline app/UserActivity.php]: Where's the `getActivityName` method?

File: resources/views/user/show.blade.php [hunk 4]
```diff
@@ -32,7 +32,26 @@ class="home-page"
...
 
         <div class="tab-content">
-            <div role="tabpanel" class="tab-pane active" id="activities">Activities here.</div>
+            <div role="tabpanel" class="tab-pane active" id="activities">
+                @if (count($user->activities))
+                    <table class="table table-striped">
+                        @foreach ($user->activities as $activity)
+                            <tr>
+                                <td>
+                                    @if ($activity->name == "followed_user")
```
Comment (nguyenthanhtung88) [inline resources/views/user/show.blade.php]: Please make `followed_user` as config

File: app/User.php [hunk 5]
```diff
@@ -79,4 +80,27 @@ public function setAvatar($avatar)
 
         $this->avatar = $filename;
     }
+
+    protected function getActivityName($event, $subject)
```
Comment (inahsanjuan) [inline app/User.php]: @nguyenthanhtung88 Here is the getActivityName method.

File: app/UserActivity.php (hunk 2 above)
Comment (inahsanjuan) [inline app/UserActivity.php]: @nguyenthanhtung88 , subject is polymorphic in user_activities table, can be morphed to any Model that will be defined from either User or Lesson. I think this requires no parameters. Am I wrong?

File: app/UserActivity.php (hunk 3 above)
Comment (inahsanjuan) [inline app/UserActivity.php]: @nguyenthanhtung88  https://github.com/framgia/fels_168/pull/9/files#diff-36f1184e0453df972dab51127e08458fR84

File: app/User.php (hunk 5 above)
Comment (nguyenthanhtung88) [inline app/User.php]: You defined in `User` class right?
And you used this method in `UserActivity` class. `UserActivity` class is extended from `Model` class, do you think it work?
I cannot see any relationship between `User` and `UserActivity` class that we can call `getActivityName` in `UserActivity` class

File: app/UserActivity.php (hunk 2 above)
Comment (nguyenthanhtung88) [inline app/UserActivity.php]: @inahsanjuan Oops my bad, it's ok 🙇

File: app/User.php (hunk 5 above)
Comment (inahsanjuan) [inline app/User.php]: @nguyenthanhtung88 , Oh sorry...I should have moved this method to `UserActivity` class. I just noticed, I have `recordActivity` method in both `User` and `UserActivity`, when it should have been in `UserActivity`. Let me fix along with other comments....

**Comment by nguyenthanhtung88** [comment]
OK

File: app/UserActivity.php [hunk 6]
```diff
@@ -3,13 +3,46 @@
...
+    {
+        $name = strtolower((new ReflectionClass($subject))->getShortName());
+        return $event . '_' . $name;
+    }
+
+    public function recordActivity($user, $event, $subject)
+    {
+        $this->name = $this->getActivityName($event, $subject);
+        $this->user_id = $user->id;
+        $this->subject_type = get_class($subject);
```
Comment (trandinhvi39) [inline app/UserActivity.php]: test

**Pull request merged**